File: resilient/resilient-annotation-processor/src/test/java/ru/tinkoff/kora/resilient/annotation/processor/aop/CircuitBreakerFallbackTests.java
package ru.tinkoff.kora.resilient.annotation.processor.aop;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import ru.tinkoff.kora.resilient.annotation.processor.aop.testdata.CircuitBreakerFallbackTarget;
import ru.tinkoff.kora.resilient.annotation.processor.aop.testdata.CircuitBreakerLifecycle;

import java.time.Duration;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class CircuitBreakerFallbackTests extends CircuitBreakerRunner {

    private CircuitBreakerFallbackTarget getService(InitializedGraph graph) {
        var values = graph.graphDraw().getNodes()
            .stream()
            .map(graph.refreshableGraph()::get)
            .toList();

        return values.stream()
            .filter(a -> a instanceof CircuitBreakerLifecycle)
            .map(a -> ((CircuitBreakerLifecycle) a).fallbackTarget())
            .findFirst().orElseThrow();
    }

    @Test
    void syncCircuitBreaker() {
        // given
        var graphDraw = createGraphDraw();
        var service = getService(graphDraw);

        // when
        try {
            service.getValueSync();
            fail("Should not happen");
        } catch (IllegalStateException e) {
            assertNotNull(e.getMessage());
        }

        // then
        final String fallback = service.getValueSync();
        assertEquals(CircuitBreakerFallbackTarget.FALLBACK, fallback);
    }

    @Test
    void monoCircuitBreaker() {
        // given
        var graphDraw = createGraphDraw();
        var service = getService(graphDraw);

        // when
        try {
            service.getValueMono().block(Duration.ofSeconds(5));
            fail("Should not happen");
        } catch (IllegalStateException e) {
            assertNotNull(e.getMessage());
        }

        // then
        final String fallback = service.getValueMono().block(Duration.ofSeconds(5));
        assertEquals(CircuitBreakerFallbackTarget.FALLBACK, fallback);
    }

    @Test
    void fluxCircuitBreaker() {
        // given
        var graphDraw = createGraphDraw();
        var service = getService(graphDraw);

        // when
        try {
            service.getValueFlux().blockFirst(Duration.ofSeconds(5));
            fail("Should not happen");
        } catch (IllegalStateException e) {
            assertNotNull(e.getMessage());
        }

        // then
        final String fallback = service.getValueFlux().blockFirst(Duration.ofSeconds(5));
        assertEquals(CircuitBreakerFallbackTarget.FALLBACK, fallback);
    }
}

File: resilient/resilient-symbol-processor/src/test/kotlin/ru/tinkoff/kora/resilient/symbol/processor/aop/CircuitBreakerFallbackTests.kt
package ru.tinkoff.kora.resilient.symbol.processor.aop

import com.google.devtools.ksp.KspExperimental
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import ru.tinkoff.kora.resilient.symbol.processor.aop.testdata.CircuitBreakerTarget
import ru.tinkoff.kora.resilient.symbol.processor.aop.testdata.CircuitBreakerFallbackTarget

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@KspExperimental
class CircuitBreakerFallbackTests : TestAppRunner() {

    @Test
    fun syncCircuitBreaker() {
        // given
        val services: Pair<CircuitBreakerTarget, CircuitBreakerFallbackTarget> = getServicesFromGraph()
        val service = services.second

        // when
        try {
            service.getValueSync()
            fail("Should not happen")
        } catch (e: IllegalStateException) {
            assertNotNull(e.message)
        }

        // then
        val fallback = service.getValueSync()
        assertEquals(CircuitBreakerFallbackTarget.FALLBACK, fallback)
    }

    @Test
    fun suspendCircuitBreaker() {
        // given
        val services: Pair<CircuitBreakerTarget, CircuitBreakerFallbackTarget> = getServicesFromGraph()
        val service = services.second

        // when
        try {
            runBlocking { service.getValueSuspend() }
            fail("Should not happen")
        } catch (e: IllegalStateException) {
            assertNotNull(e.message)
        }

        // then
        val fallback = runBlocking { service.getValueSuspend() }
        assertEquals(CircuitBreakerFallbackTarget.FALLBACK, fallback)
    }

    @Test
    fun flowCircuitBreaker() {
        // given
        val services: Pair<CircuitBreakerTarget, CircuitBreakerFallbackTarget> = getServicesFromGraph()
        val service = services.second

        // when
        try {
            runBlocking { service.getValueFLow().toList() }
            fail("Should not happen")
        } catch (e: IllegalStateException) {
            assertNotNull(e.message)
        }

        // then
        val fallback = runBlocking { service.getValueFLow().toList() }
        assertEquals(1, fallback.size)
        assertEquals(CircuitBreakerFallbackTarget.FALLBACK, fallback[0])
    }
}
